packages/components/src/Dropdown/Dropdown.tsx

```diff
@@ -0,0 +1,199 @@
+        </DropdownOption>
+      ))}
+    </DropdownOptions>
+  ) : null;
+
+  const listboxChildren = (
+    <>
+      {children}
+      {optionsList}
+    </>
+  );
+
+  return (
+    <Listbox
+      // Provide a wrapping <div> element for the dropdown. This is needed so that any props
+      // passed directly to this component have a corresponding DOM element to receive them.
+      // Otherwise we get an error.
+      as="div"
+      className={clsx(styles.dropdown, className)}
+      {...rest}
+    >
+      {listboxChildren}
+    </Listbox>
+  );
+}
+
+const DropdownTrigger = function (
+  props: PropsWithRenderProp<{ disabled: boolean; open: boolean }>,
+) {
+  const { children, className, ...rest } = props;
+  return (
+    <Listbox.Button
+      // Render as a fragment instead of the default element. We're rendering our own element in
+      // the render prop to control styling and positiong, and we don't want to end up with
+      // duplicate buttons.
+      as={React.Fragment}
+      {...rest}
+    >
+      {typeof children === "function" ? (
+        children
+      ) : (
+        <DropdownButton className={className}>{children}</DropdownButton>
+      )}
+    </Listbox.Button>
+  );
+};
+
+const DropdownOptions = function (
+  props: PropsWithRenderProp<{ open: boolean }>,
+) {
+  const { className, ...rest } = props;
+  return (
+    <Listbox.Options
+      className={clsx(styles.dropdownOptions, className)}
+      {...rest}
+    />
+  );
+};
+
+const DropdownOption = function (props: DropdownOptionProps) {
+  const { children, className, ...rest } = props;
+
+  return (
+    <Listbox.Option
+      // Render as a fragment instead of the default <li>. We're rendering our own <li> in the
+      // render prop to control active/selected styling, and we don't want to end up with duplicate
+      // <li>'s.
+      as={React.Fragment}
+      {...rest}
+    >
+      {typeof children === "function"
+        ? children
+        : ({
+            active,
+            disabled,
+            selected,
+          }: {
+            active: boolean;
+            disabled: boolean;
+            selected: boolean;
+          }) => (
+            <li
+              className={clsx(
+                styles.dropdownOption,
+                active && styles.dropdownOptionActive,
+                disabled && styles.dropdownOptionDisabled,
+                selected
+                  ? styles.dropdownOptionSelected
+                  : styles.dropdownOptionUnselected,
+                className,
+              )}
+            >
+              {selected && (
+                <span className={styles.selectedIconContainer}>
+                  <CheckRoundedIcon purpose="decorative" size="1.25rem" />
+                </span>
+              )}
+              <span className={styles.dropdownOptionContent}>{children}</span>
+            </li>
+          )}
+    </Listbox.Option>
+  );
+};
+
+Dropdown.Button = DropdownTrigger;
+Dropdown.Label = Listbox.Label;
```

Comment:
I'm getting an error for this line when trying to rebuild the package:

>src/Dropdown/Dropdown.tsx:195:1 - error TS4032: Property 'Label' of exported interface has or is using name 'LabelRenderPropArg' from private module '"/Users/drater/edu-design-system/node_modules/@headlessui/react/dist/components/listbox/listbox"'.
>
>195 Dropdown.Label = Listbox.Label;

<img width="738" alt="error output in terminal" src="https://user-images.githubusercontent.com/7761701/152447814-08ca98db-0356-48e9-8870-7a0934d154e5.png">

Reply:
That's a weird TS error I haven't seen before

Tagging @andremalan in case he's ran into something like this when converting files

Reply:
This may be basically the same as https://github.com/microsoft/TypeScript/issues/6307, since we are using the `--declaration` option (in this project's tsconfig.json).

Presumably LP doesn't have that problem because it doesn't output any declaration files.

Fix might be to re-export `LabelRenderPropArg` from Headless UI? Although that does seem unfortunate.

Reply:
I don't think we can import `LabelRenderPropArg` from Headless UI to begin with because it's not being exported 🤔 https://github.com/tailwindlabs/headlessui/blob/fbaa1ae9da8788910ce8b240851f9779630babcb/packages/%40headlessui-react/src/components/listbox/listbox.tsx#L417

Reply:
Looks like one hack around this is to add a `DropdownLabel` component that just wraps the `Listbox.Label` and sets the prop types to `any`:
```
const DropdownLabel = (props: any) => <Listbox.Label {...props} />;

Dropdown.Button = DropdownTrigger;
Dropdown.Label = DropdownLabel;
Dropdown.Option = DropdownOption;
Dropdown.Options = DropdownOptions;

export default Dropdown;
```

Reply:
> Looks like one hack around this is to add a `DropdownLabel` component that just wraps the `Listbox.Label` and sets the prop types to `any`

That could work if there's a way to replace the `any` type with types that more closely align with `Listbox.Label`. Does this give us better typing for Props?

`type Props = React.ComponentProps<typeof Listbox.Label>;`

If that works, then you could check that the [render props of Listbox.Label](https://headlessui.dev/react/listbox#listbox-label) are inferred correctly as boolean values. Or you can try checking another prop just to see if TS throws an error when you try passing an invalid value

Reply:
Using `React.ComponentProps<typeof Listbox.Label>;` unfortunately brings the issue back. 😔

Code block:
<img width="513" alt="Screen Shot 2022-02-04 at 9 15 33 AM" src="https://user-images.githubusercontent.com/7761701/152573060-251fda5d-f37f-42eb-9497-df7b5792952e.png">

Code block with type error popover visible:
<img width="846" alt="Screen Shot 2022-02-04 at 9 15 43 AM" src="https://user-images.githubusercontent.com/7761701/152573064-0bfa9a92-781a-4981-8ed9-b06c4b79dcd7.png">

Reply:
I don't really expect us to use the render props for this Label element 🤔  from the [Headless documentation](https://headlessui.dev/react/listbox#listbox-label), the only prop it expects is `as`

Could we type as
```
const DropdownLabel = (props: {
  as?: React.ElementType;
  className?: string;
}) => <Listbox.Label {...props} />;
```
?

Reply:
> I don't really expect us to use the render props for this Label element 🤔 from the [Headless documentation](https://headlessui.dev/react/listbox#listbox-label), the only prop it expects is `as`
> 
> Could we type as
> 
> ```
> const DropdownLabel = (props: {
>   as?: React.ElementType;
>   className?: string;
> }) => <Listbox.Label {...props} />;
> ```

That change produces a type error in the `.stories` file usage of the `Dropdown.Label` for some reason:

>➤ YN0000: src/Dropdown/Dropdown.stories.tsx(37,8): error TS2559: Type '{ children: string; }' has no properties in common with type 'IntrinsicAttributes & { as?: ElementType<any> | undefined; className?: string | undefined; }'.
>➤ YN0000: src/Dropdown/Dropdown.stories.tsx(62,8): error TS2559: Type '{ children: string; }' has no properties in common with type 'IntrinsicAttributes & { as?: ElementType<any> | undefined; className?: string | undefined; }'.

<img width="978" alt="type error in visual studio code editor" src="https://user-images.githubusercontent.com/7761701/152851774-97b2088c-38fc-4b51-8301-ad7320f07c72.png">

Reply:
@dierat What if you add a type for `children`?
```
const DropdownLabel = (props: {
  as?: React.ElementType;
  className?: string;
  children?: React.ReactNode
}) => <Listbox.Label {...props} />;
```

Reply:
+1 (and good catch), we could also make that prop required

Reply:
Ah of course, thanks! I think this solves the issue ❤️ ❤️ ❤️